In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb

In [2]:
heart = pd.read_csv("./processed.cleveland.data.csv")
heart.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  heart_disease  
0    3.0  0.0  6.0              0  
1    2.0  3.0  3.0              2  
2    2.0  2.0  7.0              1  
3    3.0  0.0  3.0              0  
4    1.0  0.0  3.0              0


#### Data description

- age: age in years
- sex: 1=male, 0=female
- cp: chest pain type
 - Value 1: typical angina
 - Value 2: atypical angina
 - Value 3: non-anginal pain
 - Value 4: asymptomatic
- trestbps: resting blood pressure (in mm Hg on admission to the hospital)
- chol: serum cholestoral in mg/dl
- fbs: (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)
- restecg: resting electrocardiographic results
 - Value 0: normal
 - Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
 - Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria
- thalach: maximum heart rate achieved in an exercise test
- exang: exercise induced angina (1 = yes; 0 = no)
- oldpeak: ST depression induced by exercise relative to rest
- slope: the slope of the peak exercise ST segment
 - Value 1: upsloping
 - Value 2: flat
 - Value 3: downsloping
- ca: number of major vessels (0-3) colored by flourosopy
- thal:
 - Value 3: normal
 - Value 6: fixed defect
 - Value 7: reversable defect
- heart_disease: diagnosis of heart disease (angiographic disease status)
 - Value 0: < 50% diameter narrowing
 - Value 1: > 50% diameter narrowing

"\[This field\] refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 4. Experiments with the Cleveland database have concentrated on simply attempting to distinguish presence (values 1,2,3,4) from absence (value 0)."

In [3]:
heart.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   age            303 non-null    float64
 1   sex            303 non-null    float64
 2   cp             303 non-null    float64
 3   trestbps       303 non-null    float64
 4   chol           303 non-null    float64
 5   fbs            303 non-null    float64
 6   restecg        303 non-null    float64
 7   thalach        303 non-null    float64
 8   exang          303 non-null    float64
 9   oldpeak        303 non-null    float64
 10  slope          303 non-null    float64
 11  ca             303 non-null    object 
 12  thal           303 non-null    object 
 13  heart_disease  303 non-null    int64  
dtypes: float64(11), int64(1), object(2)
memory usage: 33.3+ KB


In [4]:
heart.describe(include='all')

age         sex          cp    trestbps        chol  \
count   303.000000  303.000000  303.000000  303.000000  303.000000   
unique         NaN         NaN         NaN         NaN         NaN   
top            NaN         NaN         NaN         NaN         NaN   
freq           NaN         NaN         NaN         NaN         NaN   
mean     54.438944    0.679868    3.158416  131.689769  246.693069   
std       9.038662    0.467299    0.960126   17.599748   51.776918   
min      29.000000    0.000000    1.000000   94.000000  126.000000   
25%      48.000000    0.000000    3.000000  120.000000  211.000000   
50%      56.000000    1.000000    3.000000  130.000000  241.000000   
75%      61.000000    1.000000    4.000000  140.000000  275.000000   
max      77.000000    1.000000    4.000000  200.000000  564.000000   

               fbs     restecg     thalach       exang     oldpeak  \
count   303.000000  303.000000  303.000000  303.000000  303.000000   
unique         NaN         NaN         NaN         NaN         NaN   
top            NaN         NaN         NaN         NaN         NaN   
freq           NaN         NaN         NaN         NaN         NaN   
mean      0.148515    0.990099  149.607261    0.326733    1.039604   
std       0.356198    0.994971   22.875003    0.469794    1.161075   
min       0.000000    0.000000   71.000000    0.000000    0.000000   
25%       0.000000    0.000000  133.500000    0.000000    0.000000   
50%       0.000000    1.000000  153.000000    0.000000    0.800000   
75%       0.000000    2.000000  166.000000    1.000000    1.600000   
max       1.000000    2.000000  202.000000    1.000000    6.200000   

             slope   ca thal  heart_disease  
count   303.000000  303  303     303.000000  
unique         NaN    5    4            NaN  
top            NaN  0.0  3.0            NaN  
freq           NaN  176  166            NaN  
mean      1.600660  NaN  NaN       0.937294  
std       0.616226  NaN  NaN       1.228536  
min       1.000000  NaN  NaN       0.000000  
25%       1.000000  NaN  NaN       0.000000  
50%       2.000000  NaN  NaN       0.000000  
75%       2.000000  NaN  NaN       2.000000  
max       3.000000  NaN  NaN       4.000000

In [5]:
# fixing the weirdness with the ca and thal columns being interpreted as
# objects(strings) instead of numbers
heart.ca.unique()

array(['0.0', '3.0', '2.0', '1.0', '?'], dtype=object)

In [6]:
heart[heart.ca == '?']

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
166  52.0  1.0  3.0     138.0  223.0  0.0      0.0    169.0    0.0      0.0   
192  43.0  1.0  4.0     132.0  247.0  1.0      2.0    143.0    1.0      0.1   
287  58.0  1.0  2.0     125.0  220.0  0.0      0.0    144.0    0.0      0.4   
302  38.0  1.0  3.0     138.0  175.0  0.0      0.0    173.0    0.0      0.0   

     slope ca thal  heart_disease  
166    1.0  ?  3.0              0  
192    2.0  ?  7.0              1  
287    2.0  ?  7.0              0  
302    1.0  ?  3.0              0

In [7]:
heart[heart.thal == '?']

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
87   53.0  0.0  3.0     128.0  216.0  0.0      2.0    115.0    0.0      0.0   
266  52.0  1.0  4.0     128.0  204.0  1.0      0.0    156.0    1.0      1.0   

     slope   ca thal  heart_disease  
87     1.0  0.0    ?              0  
266    2.0  0.0    ?              2

In [8]:
heart = heart.replace('?', np.nan)
print(heart[heart.thal == '?'])
print(heart[heart.ca == '?'])

Empty DataFrame
Columns: [age, sex, cp, trestbps, chol, fbs, restecg, thalach, exang, oldpeak, slope, ca, thal, heart_disease]
Index: []
Empty DataFrame
Columns: [age, sex, cp, trestbps, chol, fbs, restecg, thalach, exang, oldpeak, slope, ca, thal, heart_disease]
Index: []


In [9]:
heart.ca = heart.ca.astype('float')
heart.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   age            303 non-null    float64
 1   sex            303 non-null    float64
 2   cp             303 non-null    float64
 3   trestbps       303 non-null    float64
 4   chol           303 non-null    float64
 5   fbs            303 non-null    float64
 6   restecg        303 non-null    float64
 7   thalach        303 non-null    float64
 8   exang          303 non-null    float64
 9   oldpeak        303 non-null    float64
 10  slope          303 non-null    float64
 11  ca             299 non-null    float64
 12  thal           301 non-null    object 
 13  heart_disease  303 non-null    int64  
dtypes: float64(12), int64(1), object(1)
memory usage: 33.3+ KB


In [10]:
heart.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   age            303 non-null    float64
 1   sex            303 non-null    float64
 2   cp             303 non-null    float64
 3   trestbps       303 non-null    float64
 4   chol           303 non-null    float64
 5   fbs            303 non-null    float64
 6   restecg        303 non-null    float64
 7   thalach        303 non-null    float64
 8   exang          303 non-null    float64
 9   oldpeak        303 non-null    float64
 10  slope          303 non-null    float64
 11  ca             299 non-null    float64
 12  thal           301 non-null    object 
 13  heart_disease  303 non-null    int64  
dtypes: float64(12), int64(1), object(1)
memory usage: 33.3+ KB


In [11]:
# sex: 1=male, 0=female
heart.sex.replace({0.0: 'female', 1.0: 'male'}, inplace=True)
heart.dtypes

age              float64
sex               object
cp               float64
trestbps         float64
chol             float64
fbs              float64
restecg          float64
thalach          float64
exang            float64
oldpeak          float64
slope            float64
ca               float64
thal              object
heart_disease      int64
dtype: object

In [12]:
#cp: chest pain type
# - Value 1: typical angina
# - Value 2: atypical angina
# - Value 3: non-anginal pain
# - Value 4: asymptomatic

heart.cp.replace({1.0: 'typical angina', 2.0: 'atypical angina',
                  3.0: 'non-anginal pain', 4.0: 'asymptomatic'}, inplace=True)
print(heart.dtypes)
heart.head()

age              float64
sex               object
cp                object
trestbps         float64
chol             float64
fbs              float64
restecg          float64
thalach          float64
exang            float64
oldpeak          float64
slope            float64
ca               float64
thal              object
heart_disease      int64
dtype: object


age     sex                cp  trestbps   chol  fbs  restecg  thalach  \
0  63.0    male    typical angina     145.0  233.0  1.0      2.0    150.0   
1  67.0    male      asymptomatic     160.0  286.0  0.0      2.0    108.0   
2  67.0    male      asymptomatic     120.0  229.0  0.0      2.0    129.0   
3  37.0    male  non-anginal pain     130.0  250.0  0.0      0.0    187.0   
4  41.0  female   atypical angina     130.0  204.0  0.0      2.0    172.0   

   exang  oldpeak  slope   ca thal  heart_disease  
0    0.0      2.3    3.0  0.0  6.0              0  
1    1.0      1.5    2.0  3.0  3.0              2  
2    1.0      2.6    2.0  2.0  7.0              1  
3    0.0      3.5    3.0  0.0  3.0              0  
4    0.0      1.4    1.0  0.0  3.0              0

In [13]:
# restecg: resting electrocardiographic results
# - Value 0: normal
# - Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
# - Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria

heart.restecg.replace({0.0: 'normal', 1.0: 'ST-T wave abnormality',
                       2.0: 'left ventricular hypertrophy'}, inplace=True)
print(heart.dtypes)
heart.head()

age              float64
sex               object
cp                object
trestbps         float64
chol             float64
fbs              float64
restecg           object
thalach          float64
exang            float64
oldpeak          float64
slope            float64
ca               float64
thal              object
heart_disease      int64
dtype: object


age     sex                cp  trestbps   chol  fbs  \
0  63.0    male    typical angina     145.0  233.0  1.0   
1  67.0    male      asymptomatic     160.0  286.0  0.0   
2  67.0    male      asymptomatic     120.0  229.0  0.0   
3  37.0    male  non-anginal pain     130.0  250.0  0.0   
4  41.0  female   atypical angina     130.0  204.0  0.0   

                        restecg  thalach  exang  oldpeak  slope   ca thal  \
0  left ventricular hypertrophy    150.0    0.0      2.3    3.0  0.0  6.0   
1  left ventricular hypertrophy    108.0    1.0      1.5    2.0  3.0  3.0   
2  left ventricular hypertrophy    129.0    1.0      2.6    2.0  2.0  7.0   
3                        normal    187.0    0.0      3.5    3.0  0.0  3.0   
4  left ventricular hypertrophy    172.0    0.0      1.4    1.0  0.0  3.0   

   heart_disease  
0              0  
1              2  
2              1  
3              0  
4              0

In [14]:
# get_dummies converts the categorical variables into indicator variable(0/1)

heart = pd.get_dummies(data=heart, columns=['restecg'])
heart.head()

age     sex                cp  trestbps   chol  fbs  thalach  exang  \
0  63.0    male    typical angina     145.0  233.0  1.0    150.0    0.0   
1  67.0    male      asymptomatic     160.0  286.0  0.0    108.0    1.0   
2  67.0    male      asymptomatic     120.0  229.0  0.0    129.0    1.0   
3  37.0    male  non-anginal pain     130.0  250.0  0.0    187.0    0.0   
4  41.0  female   atypical angina     130.0  204.0  0.0    172.0    0.0   

   oldpeak  slope   ca thal  heart_disease  restecg_ST-T wave abnormality  \
0      2.3    3.0  0.0  6.0              0                              0   
1      1.5    2.0  3.0  3.0              2                              0   
2      2.6    2.0  2.0  7.0              1                              0   
3      3.5    3.0  0.0  3.0              0                              0   
4      1.4    1.0  0.0  3.0              0                              0   

   restecg_left ventricular hypertrophy  restecg_normal  
0                                     1               0  
1                                     1               0  
2                                     1               0  
3                                     0               1  
4                                     1               0

In [15]:
# dropping columns in the dataframe

heart.drop(['restecg_ST-T wave abnormality', 'restecg_left ventricular hypertrophy'], axis=1, inplace= True)
heart.head()

age     sex                cp  trestbps   chol  fbs  thalach  exang  \
0  63.0    male    typical angina     145.0  233.0  1.0    150.0    0.0   
1  67.0    male      asymptomatic     160.0  286.0  0.0    108.0    1.0   
2  67.0    male      asymptomatic     120.0  229.0  0.0    129.0    1.0   
3  37.0    male  non-anginal pain     130.0  250.0  0.0    187.0    0.0   
4  41.0  female   atypical angina     130.0  204.0  0.0    172.0    0.0   

   oldpeak  slope   ca thal  heart_disease  restecg_normal  
0      2.3    3.0  0.0  6.0              0               0  
1      1.5    2.0  3.0  3.0              2               0  
2      2.6    2.0  2.0  7.0              1               0  
3      3.5    3.0  0.0  3.0              0               1  
4      1.4    1.0  0.0  3.0              0               0

In [16]:
# slope: the slope of the peak exercise ST segment
# - Value 1: upsloping
# - Value 2: flat
# - Value 3: downsloping
heart.slope.replace({1.0: 'upsloping', 2.0: 'flat', 3.0: 'downsloping'},
                    inplace=True)
print(heart.dtypes)
heart.head()

age               float64
sex                object
cp                 object
trestbps          float64
chol              float64
fbs               float64
thalach           float64
exang             float64
oldpeak           float64
slope              object
ca                float64
thal               object
heart_disease       int64
restecg_normal      uint8
dtype: object


age     sex                cp  trestbps   chol  fbs  thalach  exang  \
0  63.0    male    typical angina     145.0  233.0  1.0    150.0    0.0   
1  67.0    male      asymptomatic     160.0  286.0  0.0    108.0    1.0   
2  67.0    male      asymptomatic     120.0  229.0  0.0    129.0    1.0   
3  37.0    male  non-anginal pain     130.0  250.0  0.0    187.0    0.0   
4  41.0  female   atypical angina     130.0  204.0  0.0    172.0    0.0   

   oldpeak        slope   ca thal  heart_disease  restecg_normal  
0      2.3  downsloping  0.0  6.0              0               0  
1      1.5         flat  3.0  3.0              2               0  
2      2.6         flat  2.0  7.0              1               0  
3      3.5  downsloping  0.0  3.0              0               1  
4      1.4    upsloping  0.0  3.0              0               0

In [17]:
heart.slope = pd.Categorical(heart.slope, ['upsloping', 'flat', 'downsloping'],
                             ordered=True)
heart.slope.cat.codes

0      2
1      1
2      1
3      2
4      0
      ..
298    1
299    1
300    1
301    1
302    0
Length: 303, dtype: int8

In [18]:
heart.slope.cat.codes.median()

1.0

In [19]:
# thal: 3 = normal; 6 = fixed defect; 7 = reversable defect
heart.thal.replace({'3.0': 'normal', '6.0': 'fixed defect', '7.0': 'reversal defect'},
                   inplace=True )
heart.head()

age     sex                cp  trestbps   chol  fbs  thalach  exang  \
0  63.0    male    typical angina     145.0  233.0  1.0    150.0    0.0   
1  67.0    male      asymptomatic     160.0  286.0  0.0    108.0    1.0   
2  67.0    male      asymptomatic     120.0  229.0  0.0    129.0    1.0   
3  37.0    male  non-anginal pain     130.0  250.0  0.0    187.0    0.0   
4  41.0  female   atypical angina     130.0  204.0  0.0    172.0    0.0   

   oldpeak        slope   ca             thal  heart_disease  restecg_normal  
0      2.3  downsloping  0.0     fixed defect              0               0  
1      1.5         flat  3.0           normal              2               0  
2      2.6         flat  2.0  reversal defect              1               0  
3      3.5  downsloping  0.0           normal              0               1  
4      1.4    upsloping  0.0           normal              0               0

- heart_disease: diagnosis of heart disease (angiographic disease status)
 - Value 0: < 50% diameter narrowing
 - Value 1: > 50% diameter narrowing

"\[This field\] refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 4. Experiments with the Cleveland database have concentrated on simply attempting to distinguish presence (values 1,2,3,4) from absence (value 0)."

In [20]:
np.where(heart.heart_disease == 0, 'absence', 'presence')[1:10]

array(['presence', 'presence', 'absence', 'absence', 'absence',
       'presence', 'absence', 'presence', 'presence'], dtype='<U8')

In [21]:
heart.heart_disease[1:10]

1    2
2    1
3    0
4    0
5    0
6    3
7    0
8    2
9    1
Name: heart_disease, dtype: int64

In [22]:
heart.heart_disease = np.where(heart.heart_disease == 0, 'absence', 'presence')
heart.head()

age     sex                cp  trestbps   chol  fbs  thalach  exang  \
0  63.0    male    typical angina     145.0  233.0  1.0    150.0    0.0   
1  67.0    male      asymptomatic     160.0  286.0  0.0    108.0    1.0   
2  67.0    male      asymptomatic     120.0  229.0  0.0    129.0    1.0   
3  37.0    male  non-anginal pain     130.0  250.0  0.0    187.0    0.0   
4  41.0  female   atypical angina     130.0  204.0  0.0    172.0    0.0   

   oldpeak        slope   ca             thal heart_disease  restecg_normal  
0      2.3  downsloping  0.0     fixed defect       absence               0  
1      1.5         flat  3.0           normal      presence               0  
2      2.6         flat  2.0  reversal defect      presence               0  
3      3.5  downsloping  0.0           normal       absence               1  
4      1.4    upsloping  0.0           normal       absence               0

In [23]:
heart.describe(include='all')


age   sex            cp    trestbps        chol         fbs  \
count   303.000000   303           303  303.000000  303.000000  303.000000   
unique         NaN     2             4         NaN         NaN         NaN   
top            NaN  male  asymptomatic         NaN         NaN         NaN   
freq           NaN   206           144         NaN         NaN         NaN   
mean     54.438944   NaN           NaN  131.689769  246.693069    0.148515   
std       9.038662   NaN           NaN   17.599748   51.776918    0.356198   
min      29.000000   NaN           NaN   94.000000  126.000000    0.000000   
25%      48.000000   NaN           NaN  120.000000  211.000000    0.000000   
50%      56.000000   NaN           NaN  130.000000  241.000000    0.000000   
75%      61.000000   NaN           NaN  140.000000  275.000000    0.000000   
max      77.000000   NaN           NaN  200.000000  564.000000    1.000000   

           thalach       exang     oldpeak      slope          ca    thal  \
count   303.000000  303.000000  303.000000        303  299.000000     301   
unique         NaN         NaN         NaN          3         NaN       3   
top            NaN         NaN         NaN  upsloping         NaN  normal   
freq           NaN         NaN         NaN        142         NaN     166   
mean    149.607261    0.326733    1.039604        NaN    0.672241     NaN   
std      22.875003    0.469794    1.161075        NaN    0.937438     NaN   
min      71.000000    0.000000    0.000000        NaN    0.000000     NaN   
25%     133.500000    0.000000    0.000000        NaN    0.000000     NaN   
50%     153.000000    0.000000    0.800000        NaN    0.000000     NaN   
75%     166.000000    1.000000    1.600000        NaN    1.000000     NaN   
max     202.000000    1.000000    6.200000        NaN    3.000000     NaN   

       heart_disease  restecg_normal  
count            303      303.000000  
unique             2             NaN  
top          absence             NaN  
freq             164             NaN  
mean             NaN        0.498350  
std              NaN        0.500824  
min              NaN        0.000000  
25%              NaN        0.000000  
50%              NaN        0.000000  
75%              NaN        1.000000  
max              NaN        1.000000